src/fastfield/multivalued/writer.rs

```diff
@@ -94,7 +93,35 @@ impl MultiValuedFastFieldWriter {
         self.vals.extend_from_slice(vals);
         doc
     }
+    /// Returns an iterator over values per docid in ascending docid order.
+    ///
+    /// Normally the order is simply iterating self.docid_index.
+    /// With docid_map it accounts for the new mapping, returning values in the order of the
+    /// new docids.
+    fn get_ordered_values<'a: 'b, 'b>(
+        &'a self,
+        docid_map: Option<&'b DocidMapping>,
+    ) -> impl Iterator<Item = &'b [u64]> {
+        let docid_iter = if let Some(docid_map) = docid_map {
+            Box::new(docid_map.iter_old_docids().cloned()) as Box<dyn Iterator<Item = u32>>
+        } else {
+            Box::new(self.doc_index.iter().enumerate().map(|el| el.0 as u32))
+                as Box<dyn Iterator<Item = u32>>
+        };
+        docid_iter.map(move |docid| self.get_values_for_docid(docid))
+    }
 
+    /// returns all values for a docids
+    fn get_values_for_docid(&self, docid: u32) -> &[u64] {
+        let start_pos = self.doc_index[docid as usize] as usize;
+        let end_pos = if docid as usize + 1 == self.doc_index.len() {
```

Comment:
i recently learn (was it in= your PR or maybe @evanxg852000 ?) that you can also do:

```
self.doc_index.get(docid as usize + 1).cloned().unwrap_or(self.vals.len())
```

Reply:
yes I used something similar recently.